### PR TITLE
 Add AES encryption write support

### PR DESCRIPTION
**Pull request by kyrias**
Replacement for zip-rs/zip-old#363.

---

By the way, it would be nice if this repo had its "forked from zip-rs/zip-old" status removed, because currently it's rather inconvenient to interact with this repository using the GitHub official CLI, since it doesn't understand that this is the current upstream repo now.

Though it also feels rather strange for this repo to be called zip2 when the crate is still called `zip` and is not at version 2.

### CODE REVIEW

**Comment by Pr0methean** [comment]
If I renamed it `zip`, it would break some links to the old repo that are currently redirects. As for detaching the fork, according to https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/detaching-a-fork that would mean resetting the issues and PRs again.

**Comment by Pr0methean** [comment]
Please sign your commits.

File: src/aes.rs [hunk 1]
```diff
@@ -184,3 +206,196 @@ impl<R: Read> AesReaderValid<R> {
...
+
+        // Reset cursor position to the beginning.
+        buf.set_position(0);
+
+        {
+            let compressed_length = buf.get_ref().len() as u64;
+            let mut reader =
+                match AesReader::new(&mut buf, aes_mode, compressed_length).validate(&password)? {
+                    None => {
+                        return Err(io::Error::new(
```
Comment (Pr0methean) [inline src/aes.rs:334]: Use ZipError::InvalidPassword instead; it was added in https://github.com/zip-rs/zip2/commit/ece098d39332f97b04e508222d0cc44ae3c1e5c4.

**Review by Pr0methean — CHANGES_REQUESTED** [review]

**Comment by kyrias** [comment]
Missing `Arbitrary` implementations added.  (Once the `arbitrary` crate gets a new release the `EncryptWith` implementation will be able to be derived.)

**Review by Pr0methean — DISMISSED** [review]

**Comment by Pr0methean** [comment]
This PR failed the `fuzz_write` [test](https://github.com/zip-rs/zip2/actions/runs/8941393103/job/24566139784?pr=80) with
```
called `Result::unwrap()` on an `Err` value: InvalidArchive("Invalid zip64 locator digital signature header")
```
Please fix.

**Comment by kyrias** [comment]
The issue is actually with adding a _directory_ with encryption enabled.  Need to resurrect my Zip packet dumper and compare, but presumably WinZip doesn't try to encrypt directories, since that would make no sense.  :sweat_smile:

**Comment by Pr0methean** [comment]
Your latest commit doesn't seem to have a GPG signature.

**Comment by Pr0methean** [comment]
@kyrias Is it supposed to be possible to write a symlink with encryption? I'm under the impression that a symlink is just a normal file with one or two extra metadata bits set, and where the plaintext is the path it resolves to. But tonight the fuzz_write test (to which I'd just added `merge_archive()`) failed on this test case:

```rust
        FuzzTestCase {
            comment: [],
            operations: [
                (
                    FileOperation {
                        basic: MergeWithOtherFile {
                            first: FileOperation {
                                basic: WriteSymlinkWithTarget {
                                    target: "",
                                    options: FileOptions {
                                        compression_method: Stored,
                                        compression_level: None,
                                        last_modified_time: DateTime {
                                            year: 1981,
                                            month: 1,
                                            day: 5,
                                            hour: 0,
                                            minute: 0,
                                            second: 0,
                                        },
                                        permissions: None,
                                        large_file: false,
                                        encrypt_with: None,
                                        extended_options: ExtendedFileOptions {
                                            extra_data: [],
                                            central_extra_data: [],
                                        },
                                        alignment: 46036,
                                        zopfli_buffer_size: None,
                                    },
                                },
                                path: "\0",
                                reopen: ViaFinishIntoReadable,
                            },
                            rest: [
                                FileOperation {
                                    basic: WriteSymlinkWithTarget {
                                        target: "",
                                        options: FileOptions {
                                            compression_method: Deflated,
                                            compression_level: Some(
                                                5,
                                            ),
                                            last_modified_time: DateTime {
                                                year: 2107,
                                                month: 4,
                                                day: 8,
                                                hour: 15,
                                                minute: 15,
                                                second: 11,
                                            },
                                            permissions: Some(
                                                4294967295,
                                            ),
                                            large_file: true,
                                            encrypt_with: Some(
                                                Aes {
                                                    mode: Aes128,
                                                    password: "\0\0",
                                                },
                                            ),
                                            extended_options: ExtendedFileOptions {
                                                extra_data: [],
                                                central_extra_data: [],
                                            },
                                            alignment: 46036,
                                            zopfli_buffer_size: None,
                                        },
                                    },
                                    path: "",
                                    reopen: ViaFinishIntoReadable,
                                },
                                FileOperation {
                                    basic: WriteNormalFile {
                                        contents: [],
                                        options: FileOptions {
                                            compression_method: Stored,
                                            compression_level: None,
                                            last_modified_time: DateTime {
                                                year: 1980,
                                                month: 1,
                                                day: 1,
                                                hour: 0,
                                                minute: 0,
                                                second: 0,
                                            },
                                            permissions: None,
                                            large_file: false,
                                            encrypt_with: None,
                                            extended_options: ExtendedFileOptions {
                                                extra_data: [],
                                                central_extra_data: [],
                                            },
                                            alignment: 0,
                                            zopfli_buffer_size: None,
                                        },
                                    },
                                    path: "",
                                    reopen: DoNotReopen,
                                },
                            ],
                        },
                        path: "",
                        reopen: DoNotReopen,
                    },
                    false,
                ),
            ],
            flush_on_finish_file: false,
        }
```

**Comment by kyrias** [comment]
@Pr0methean Ah, no, `add_symlink` should also have an `encrypt_with` override like `add_directory` does.  Will you add that or do you want me to?

**Comment by Pr0methean** [comment]
If you can do so, I'll focus on the existing crash in the fuzz test.

**Pull request merged**